platform/linux-generic/test/validation/api/ml/ml_linux.c

```diff
@@ -0,0 +1,1259 @@
+ * in y. The first dimension of input and output tensors represent batch size,
+ * thus it must be the same for all tensors here. The dynamic dimension size
+ * in the output tensor here can be deduced from the given batch size, thus no
+ * need for the implementation to fill it.
+ */
+#define NUM_COLUMN 3
+#define MAX_BATCH_SIZE 4
+#define SIZE (NUM_COLUMN * MAX_BATCH_SIZE * sizeof(double))
+static void run_model_batch_add(void)
+{
+	int ret;
+	odp_ml_data_t data;
+	odp_ml_model_t model;
+	odp_ml_data_seg_t input_segs[SIZE * 2];
+	odp_ml_data_seg_t output_segs[SIZE];
+	odp_ml_run_result_t result;
+	odp_ml_run_param_t run_param;
+	odp_ml_model_param_t model_param;
+
+	double y[12];
+	double y_expected[12];
+	uint32_t batch_size = MAX_BATCH_SIZE;
+	double x1[12] = {97, 47, 62, 19, 93, 59, 67, 42, 28, 55, 46, 31};
+	double x2[12] = {81, 56, 27, 4, 69, 12, 91, 98, 23, 90, 52, 64};
+
+	for (int i = 0; i < 12; i++)
+		y_expected[i] = x1[i] + x2[i];
+
+	odp_ml_model_param_init(&model_param);
+
+	odp_ml_data_format_t input_format[2] = {
+		{
+			.data_type = ODP_ML_DATA_TYPE_FP64,
+			.data_type_size = 8,
+			.shape.type = ODP_ML_SHAPE_BATCH,
+			.shape.num_dim = 2,
+			.shape.dim = {ODP_ML_DIM_DYNAMIC, NUM_COLUMN},
+			.shape.dim_max = {MAX_BATCH_SIZE, NUM_COLUMN}
+		},
+		{
+			.data_type = ODP_ML_DATA_TYPE_FP64,
+			.data_type_size = 8,
+			.shape.type = ODP_ML_SHAPE_BATCH,
+			.shape.num_dim = 2,
+			.shape.dim = {ODP_ML_DIM_DYNAMIC, NUM_COLUMN},
+			.shape.dim_max = {MAX_BATCH_SIZE, NUM_COLUMN}
+		}
+	};
+
+	model_param.extra_info.num_inputs = 2;
+	model_param.extra_info.input_format = input_format;
+
+	/* Verify model info about matrix_mul.onnx */
+	if (fill_model_param("batch_add.onnx", &model_param))
+		return;
+
+	model = odp_ml_model_create("batch_add", &model_param);
+	free(model_param.model);
+	CU_ASSERT_NOT_EQUAL(model, ODP_ML_MODEL_INVALID);
+	if (!model)
+		return;
+
+	if (odp_ml_model_load(model, NULL)) {
+		CU_ASSERT_EQUAL(odp_ml_model_destroy(model), 0);
+		return;
+	}
+
+	odp_ml_model_print(model);
+
+	/* Prepare parameters for running inference */
+	odp_ml_run_param_init(&run_param);
+	run_param.result = &result;
+
+	data.num_input_seg = 2;
+	data.input_seg = input_segs;
+	input_segs[0].addr = x1;
+	input_segs[1].addr = x2;
+
+	data.num_output_seg = 1;
+	data.output_seg = output_segs;
+	output_segs[0].size = sizeof(y);
+	output_segs[0].addr = y;
+
+	/* Test different batch sizes */
+	for (int i = 0; i < MAX_BATCH_SIZE; i++) {
+		run_param.batch_size = batch_size;
+		input_segs[0].size = sizeof(double) * NUM_COLUMN * batch_size;
+		input_segs[1].size = sizeof(double) * NUM_COLUMN * batch_size;
+		ret = odp_ml_run(model, &data, &run_param);
+		CU_ASSERT_EQUAL(ret, 1);
+		if (ret != 1)
+			goto fail;
+
+		for (uint32_t j = 0; j < batch_size * NUM_COLUMN; j++)
+			CU_ASSERT_EQUAL(y[j], y_expected[j]);
+
+		batch_size--;
+	}
+
+	/* Result can be NULL, when output batch size is not needed */
+	run_param.result = NULL;
```

Comment:
The comment refers to dynamic dimensions on output. Result does not have output batch size. Could simply say: "Test also without run results"